lib/rex/proto/ldap/auth.rb

```diff
@@ -0,0 +1,353 @@
+require 'net/ldap'
+require 'net/ldap/dn'
+
+module Rex
+  module Proto
+    module LDAP
+      class Auth
+        NTLM_CONST = Rex::Proto::NTLM::Constants
+        NTLM_CRYPT = Rex::Proto::NTLM::Crypt
+        MESSAGE = Rex::Proto::NTLM::Message
+
+        #
+        # Initialize the required variables
+        #
+        # @param challenge [String] NTLM Server Challenge
+        # @param domain    [String] Domain value used in NTLM
+        # @param server    [String] Server value used in NTLM
+        # @param dnsname   [String] DNS Name value used in NTLM
+        # @param dnsdomain [String] DNS Domain value used in NTLM
+        def initialize(challenge = Rex::Text.rand_text_alphanumeric(16), domain  = 'DOMAIN',
+                       server = 'SERVER', dnsname = 'server', dnsdomain = 'example.com')
+          @domain = domain
+          @server = server
+          @dnsname = dnsname
+          @dnsdomain = dnsdomain
+          @challenge = [challenge].pack('H*')
+        end
+
+        #
+        # Process the incoming LDAP login requests from clients
+        #
+        # @param user_login [OpennStruct] User login information
+        #
+        # @return auth_info [Hash] Processed authentication information
+        def process_login_request(user_login)
+          auth_info = {}
+
+          if user_login.name.empty? && user_login.authentication.empty? # Anonymous
+            auth_info = handle_anonymous_request(user_login, auth_info)
+          elsif !user_login.name.empty? # Simple
+            auth_info = handle_simple_request(user_login, auth_info)
+          elsif user_login.authentication[0] == 'GSS-SPNEGO' # SASL especially SPNEGO
+            auth_info = handle_sasl_request(user_login, auth_info)
+          else
+            auth_info[:result_code] = Net::LDAP::ResultCodeUnwillingToPerform
```

Comment:
The conditional here is for what auth methods are supported so the error response should reflect the lack of support.

```suggestion
            auth_info[:result_code] = Net::LDAP::ResultCodeAuthMethodNotSupported 
```